### ISSUE TITLE
[FR] Fire event on change

### ISSUE THREAD

**Issue by Sineos** [fellinga/node-red-contrib-ui-time-scheduler#75]
I'm using events to schedule my room heating. If the event value is changed, I need to wait until the event is due before the change actually is applied.

It would be great if you added an option to fire the event when changed, provided the change is within the current time range and selected days.

I current work around this with a function node, which has the same number of outputs like ui-time-scheduler and this hacky code:
```js
const CurTS = new Date();
const TempTS = new Date();
// Return 6 for Sunday and 0 for Monday
const CurWeekDay = (CurTS.getDay() + 6) % 7;
// Initialize an array with the number of outputs. Use "null"
// to avoid setting one if skipped in the compare
const OutArray = [null, null, null, null, null, null, null, null, null, null, null, null, null, null, null];
const data = JSON.parse(msg.payload);

data.timers.forEach((obj) => {
  const temp = new Date(obj.starttime);
  const CompareTS = TempTS.setHours(temp.getHours(), temp.getMinutes(), 0);

  if (CurTS.getTime() > CompareTS && Number(obj.days[CurWeekDay]) === 1) {
    OutArray[Number(obj.output)] = {payload: Number(obj.event)};
  }
});
return OutArray; 
```
Basically using the JSON status output on port 0 to create another message. This is a weak solution since it again forces all values, so an improvement would be highly appreciated. Unfortunately my JS skills are too basic for your node to send a PR.

**Comment by fellinga** [comment]
Hi Sineos, 

I am a little confused about "if it's within time range". You said you use the event mode, and the event mode is literally a single point in time. So you can't really figure out if it's within some range. Comparing it against new Date() is problematic as there could be multiple events after the one currently evaluated.

What is your definition of "within range" and would your scenario maybe work better if the default mode (from/to) would support something like custom payloads?

**Comment by Sineos** [comment]
Thanks for coming back to this and sorry for not being clear in the first place.

A typical entry looks like:
![grafik](https://user-images.githubusercontent.com/4459312/163130514-c3b0a71a-5143-4490-9c1d-7610d1e1adc3.png)

* At 04:30 my thermostat receives the 21°C event
* If I now change the target at 10:30 to say 22°C, I need to wait until next day 04:30 until it becomes effective

What the function node does:
* Takes the JSON output on port 0 that is generated on change
* Compares if the (Current Local Time) is greater than the (Event Time): 10:30 is greater than 04:30 but below the second time 22:00 --> This is what I meant "within range"
* Compares if the (days array) == 1 for (current week day)
* Outputs the (event value) on the respective port of the function node and thus sets the new (event value) to the thermostat

> would your scenario maybe work better if the default mode (from/to) would support something like custom payloads?

This would be a nice solution as well, but would need:
* Custom payload to be set in the UI
* Also "fire once" on change or it will be the same as explained above

Many thanks for considering.

**Comment by fellinga** [comment]
The approach to compare two entries in the event mode is problematic. First of all, the reason why the event mode was created is to schedule single events at a single time. Comparing one event to other events in the list is therefore something that does not feel right. Also, if there is no second element in the list then things would become a little confusing, since I would have to choose to update always, or never.

The default mode works a little different. It already updates regularly (once per minute, customizable) and also on change. That would work well with custom payloads, but not quite. Updating regularly currently means it sends an on signal (true) or an off signal (false) whether one of the schedules are currently active or not. I am not quite sure what I would need to send outside of schedules times then.. false, some fixed payload, nothing?

**Comment by Sineos** [comment]
Many thanks for your support. Really appreciated.

> Comparing one event to other events in the list is therefore something that does not feel right. Also, if there is no second element in the list then things would become a little confusing, since I would have to choose to update always, or never.

I'm not really happy with this either, this is why I opened this request.

> I know that this is not a direct solution to what you have been asking for but with the additional controls you can easily readjust the temperature after a timer change and it also provides an additional nice feature to your flow.

Nice one 👍 
Unless I'm now missing something, this works perfectly for one event. 
My heating control works on a room basis, so I currently have 15 rooms with different events, according to its usage profiles, i.e. school time, weekend etc.
In addition, I'm storing away the JSON output in an flow context array, so I can create different "profiles", e.g. a temp setting for all rooms in autumn, spring and winter.

> Updating regularly currently means it sends an on signal (true) or an off signal (false) whether one of the schedules are currently active or not. I am not quite sure what I would need to send outside of schedules times then.. false, some fixed payload, nothing?

For my usage scenario, a combination of "default mode" and "event mode" would be perfect. I.e. being able to provide a custom value for each timer.

**Comment by fellinga** [comment]
Going to close this issue since a combination of "default mode" and "event mode" is not planned.
